### PR TITLE
Implementing the use of TrainingDataImporter in rasa test 

### PR DESCRIPTION
**Pull request by Imod7**
Implementing the 2nd Part of the changes described in [Issue 5986](https://github.com/RasaHQ/rasa/issues/5986) related to `run_evaluation`.
_The current PR is a continuation of this one [#7539](https://github.com/RasaHQ/rasa/pull/7539) - in the 7539 I had issues with the CLA certificate and then squash and rebase became very difficult because of merged commits so I decided to create a new one, the current one._

**Proposed changes**:
- Loading `test data` using `TrainingDataImporter` in `run_evaluation` function.
- Changed the `Optional` parameters in `load_from_dict` of TrainingDataImporter to make it work with the data we have in `run_evaluation`.
- Instead of having 1 function, the `run_nlu_test` now we have 2 functions described below : 
   1. `run_nlu_test` that runs `run_nlu_test_async` in the event loop.
   2. `run_nlu_test_async`
- Passing directly the specific arguments to `run_nlu_test_async` instead of `argparse.Namespace` thus making it more reusable.
- Made async the following functions : 
  1. `run_evaluation`
  2. `compare_nlu`
  3. `test_nlu`
  4. `compare_nlu_models`
  5. `train_nlu_async` 
  6. `test_nlu_comparison`
- Making `train_nlu_async` function public so we can call it from other modules.

### CODE REVIEW

File: rasa/train.py
Comment (wochinge) [inline rasa/train.py:700]: why was this change required?

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:174]: As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.

What do you think? 

That's a short and not complete example of what I mean:

```python
def run_nlu_test(args: Argparse) -> None:
   rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))

async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
    # implementation
```

**Review by wochinge — COMMENTED** [review]
Great iteration! 💯 Left a few comments

**Comment by wochinge** [comment]
@Imod7 I don't understand this part in your proposed changes, what do you mean?

> Making async a list of functions and then adding the corresponding await statements when they are called.

File: rasa/train.py
Comment (Imod7) [reply rasa/train.py:700]: Because I had errors of type `RuntimeError: this event loop is already running.`. 

I just checked it again and for example for these tests  : 
```
FAILED tests/test_train.py::test_train_nlu_temp_files - RuntimeError: this event loop is already running.
FAILED tests/test_train.py::test_train_nlu_wrong_format_error_message - RuntimeError: this event loop is alre...
FAILED tests/test_train.py::test_train_nlu_with_responses_no_domain_warns - RuntimeError: this event loop is ...
FAILED tests/test_train.py::test_train_nlu_with_responses_and_domain_no_warns - RuntimeError: this event loop...
FAILED tests/test_train.py::test_train_nlu_no_nlu_file_error_message - RuntimeError: this event loop is alrea...
FAILED tests/test_train.py::test_train_nlu_autoconfig - RuntimeError: this event loop is already running.
```
I get the above errors when I have the `run_in_loop`. When I change to `await` the tests pass.

Comment (wochinge) [reply rasa/train.py:700]: Mhm, interesting 🤔  Why did this error happen? What's the stack trace? Could it be that [this](https://github.com/RasaHQ/rasa/blob/a7387f445604be5d123b776a2f4d1521990b8453/rasa/nlu/test.py#L1900) should rather call the `async` version of `train_nlu` (which is `_train_nlu_async`)?

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py:174]: > @Imod7 I don't understand this part in your proposed changes, what do you mean?
> 
> > Making async a list of functions and then adding the corresponding await statements when they are called.

@wochinge For someone who doesn't know the issue at all, I just thought it would be good to mention that I added `async` and `await` statements in the definition and the calling of a list of functions. Like a summary of some of the changes I did instead of mentioning the name of the functions one by one and where those changes took place. They can also look at the files that were changed but I thought it is helpful for someone to get a quick idea before checking the files.

Comment (Imod7) [reply rasa/cli/test.py:174]: > As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.
> 
> What do you think?
> 
> That's a short and not complete example of what I mean:
> 
> ```python
> def run_nlu_test(args: Argparse) -> None:
>    rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))
> 
> async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
>     # implementation
> ```

@wochinge Yes! I wanted to ask you about this one because I didn't know how to change it, like what type of arguments to pass. So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)? Can I not use something like **kwargs? Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

Comment (wochinge) [reply rasa/cli/test.py:174]: > So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)?

Yes 👍🏻 This has the following advantage:
* it's clear to the caller which arguments are expected
* We can type annotate each parameter which then our CI can check which in turn makes our code more robust

> Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

I'd argue that it is the responsibility of the person adding the extra option to make sure it's used correctly. If you'd use `kwargs` then we'd still have to extra the value at some point in the code. This means the person adding the extra option would have to change other code anyway. 
Or maybe from a different perspective: If all our code just passes around some dictionaries then we end up with a giant black box. When you want to understand whether / how something is at some point a provided parameters in such a codebase you'd have to understand a lot of prior code. Does that make sense?

> didn't know how to change it, like what type of arguments to pass.

To see which parameters you need, you can look at the function and see where we call `args.<some value name>`. To see which type they should have you can look at the functions which our function calls and see how these values are annotated there. Consider the following example:

```python
def foo(args: argparse.Namespace) -> None:
   my_param = args.test
   bar(my_param)
   bar2(args.output_directory)
   
def bar(value: Text, some_value: Optional[float] = None) -> None:
   pass

def bar(dir: Optional[Text] = None) -> None:
   pass
```

By looking at the implementation of `foo` you can see that there are two CLI parameters `test` and `output_directory`. After looking at the type annotation in `bar` and `bar2` you can see that they have type `Text` and type `Optional[Text]`. The explicit version of `foo` would hence look like:

```python
def foo_explicit(my_param: Text, output_directory: Optional[Text) -> None:
   bar(my_param)
   bar2(output_directory)
```

Comment (Imod7) [reply rasa/cli/test.py:195]: @wochinge Changed the parameters of `run_nlu_test_async` so it's called with the specific arguments and not with `argparse.Namespace` anymore. 
Some thoughts : 
- Maybe passing `vars(args)` as a first argument (which goes to `all_args: Optional[Dict]`) is not a good idea and I should : 
   - construct the `Dict` inside the `run_nlu_test_async` function by using all the available parameters or
   - construct the `Dict` inside the `run_nlu_test`, then passing it as a Dict argument in `run_nlu_test_async`
That way I avoid any black boxes that you mention in your comments.
- Default values are set in the parameters. But every time someone decides to change the default values in general, then someone also has to change the default values in `run_nlu_test_async`. Is this a good practice? Maybe these default values are set in some global variables somewhere? 
- I added the `docstring` from my understanding of the use of each function parameter.

File: rasa/train.py
Comment (Imod7) [reply rasa/train.py:700]: @wochinge Hey Tobias! 
Here are the error messages [trace.txt](https://github.com/RasaHQ/rasa/files/5768791/trace.txt)
I can check and try to call `_train_nlu_async`

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:120]: I'd suggest to make this the last argument (similar as `additional_arguments` in `perform_nlu_cross_validation` or `test_nlu`

File: rasa/cli/test.py
Comment (wochinge) [reply rasa/cli/test.py:195]: > Maybe passing vars(args) as a first argument (which goes to all_args: Optional[Dict]) is not a good idea and I should 

You have a very good point there. This whole `vars` thing is in my opinion not clean at all. I'm a bit worried though that this PR is getting to big, so I'd actually keep your solution for now. If you want you can tackle this in a follow-up PR?

> Default values are set in the parameters. 

I'd omit them for now as they shouldn't be required. 

> I added the docstring from my understanding of the use of each function parameter.
 Write
 
 Sounds good. Will review them in the final review 👍🏻

File: rasa/train.py
Comment (wochinge) [reply rasa/train.py:700]: I double checked for `test_train_nlu_temp_files` and if you don't make the test function `async`, you can use `rasa.utils.common.run_in_loop` as before. 

The issue is that when you define a test function via

```python
async def test_something():
   pass
```
it will start a event loop. When your test code starts another event loop via `rasa.utils.common.run_in_loop` you get the error above. As you don't need the `async` for anything in this test, you can remove it safely.

File: rasa/__main__.py
Comment (wochinge) [inline rasa/__main__.py:120]: this is no longer needed, no?

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:120]: according to your docstring it should be this, no?
```suggestion
    config: Optional[Union[Text, List[Text]],
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:133]: can please wrap the docstring so that it doesn't exceed our 88 char limit?

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:137]:
```suggestion
        models_path: Path to a trained Rasa model..
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:139]: I suggest not repeating the types in docstrings as we have the type annotations for this.
```suggestion
        cross_validation: indicates if it should test the model using cross validation or not.
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:141]:
```suggestion
        runs: Number of comparison runs to make.
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:142]: I suggest not repeating the types in docstrings as we have the type annotations for this.

I suggest not repeating the types in docstrings as we have the type annotations for this.


```suggestion
        no_errors: indicates if incorrect predictions should be written to a file or not.
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:125]: As `compare_nlu_models` only accepts a `List[int]` this should also only be able to accept `List[int]`
```suggestion
    percentages:List[int],
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:124]: As the our `argparse` config specifies a default, `None` can't happen
```suggestion
    cross_validation:bool,
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:126]: As the our `argparse` config specifies a default, `None` can't happen

```suggestion
    runs: int,
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:128]: How can this parameter become `None` in the existing calls?
```suggestion
    all_args: Dict[Text, Any],
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:196]: can you please update the docstring?

File: rasa/test.py
Comment (wochinge) [inline rasa/test.py:115]: I think it's better to describe what and why things are happening in a docstring instead of describing how things are done. 
```suggestion
    """Tests a trained Core model against a set of test stories."""
```

**Review by wochinge — COMMENTED** [review]
Wuhuu, we're nearly there 💯 Awesome work with disentangling all the `async` things 🚀

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py:196]: Maybe this one 
```
"""Runs NLU tests by calling run_nlu_test_async in the event loop.

    Args:
        args: all arguments that were given and parsed.
    """
```

Comment (wochinge) [reply rasa/cli/test.py:196]: The `args` part feels a bit long. How about `the parsed CLI arguments for 'rasa test nlu'`?

**Comment by Imod7** [comment]
I think I did all the last changes that you mentioned and I also removed `Optional` from these parameters in `run_nlu_test_async` :
```
data_path: Text,
models_path: Text,
output_dir: Text,
no_errors: bool,
```
because as you showed yesterday I checked the functions related to `set_test_nlu_arguments` and : 
- in `add_nlu_data_param` it adds default nlu/data path `data`
- in `add_model_param` it adds the default models path/directory `models`
- in `add_test_nlu_argument_group`  it sets the default output directory `results`
- in `add_errors_success_params` it sets default no_errors `False`

Changed the docstring again of `run_nlu_test_async`  and `run_nlu_test` (edited)

File: rasa/__main__.py
Comment (wochinge) [inline rasa/__main__.py:6]: this import is unused

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:122]: that can be `None`, no?
```suggestion
    models_path: Optional[Text],
```

File: rasa/cli/test.py
Comment (wochinge) [inline rasa/cli/test.py:123]: this can also be `None`, no?
```suggestion
    output_dir: Optional[Text],
```

**Review by wochinge — COMMENTED** [review]
Two more comments. Can you please also update the pull request description and convert your PR from a draft to a actual pull request?

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py:123]: Same as above

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py:122]: So I thought I should remove the `Optional` because I saw that it always gets a default value so it will never be `None` as long as it is run through the command line. 
`_StoreAction(option_strings=['-m', '--model'], dest='model', nargs=None, const=None, default='models'`

Comment (wochinge) [reply rasa/cli/test.py:122]: Very good point! That can indeed not be `None`. Sorry for the confusion!

File: rasa/cli/test.py
Comment (wochinge) [reply rasa/cli/test.py:123]: 👍🏻

File: rasa/__main__.py
Comment (wochinge) [reply rasa/__main__.py:6]: @Imod7 This is still here?

Comment (Imod7) [reply rasa/__main__.py:6]: Yes, I was waiting for feedback on the other 2 comments and then I would push. I will do a final check and push now.

Comment (wochinge) [reply rasa/__main__.py:6]: 👍🏻 Could please also resolve the comments further up in the PR which you've worked in? Makes it easier to see which changes weren't addressed yet

Comment (Imod7) [reply rasa/__main__.py:6]: Yes! I will change the description of the PR and convert it from draft also.

**Comment by wochinge** [comment]
Just re-request review once you're done and I have a final look + approve 🎉

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: I removed all default values from the parameters of run_nlu_test_async

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: Moved `all_args` as the last parameter in `run_nlu_test_async`.

File: rasa/__main__.py
Comment (Imod7) [reply rasa/__main__.py]: True! I removed the coroutine check and it calls directly `cmdline_arguments.func(cmdline_arguments)` as before.

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: True! I changed it as you suggested.

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: I did! I also changed the setting in Pycharm so that the wrapping line is at 88 characters and this hopefully does not happen again.

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: Thank you for the suggestion. Changed it!

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: True! I removed all types from doc strings.

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: Changed it !

File: rasa/cli/test.py
Comment (wochinge) [reply rasa/cli/test.py]: slightly related and nice to know: https://en.wikipedia.org/wiki/Hungarian_notation

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: Changed it!

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: True! Changed it!

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: True! Changed it!

File: rasa/cli/test.py
Comment (Imod7) [reply rasa/cli/test.py]: I adjusted it!

File: rasa/test.py
Comment (Imod7) [reply rasa/test.py]: Changed it! I will keep this advice also for all the future docstrings.

**Review by wochinge — APPROVED** [review]
Congrats to your first Rasa Open Source change 🚀 

Two notes:
* It's (in my opinion) ok to stay a little bit more high level in your pull request description. I can see the details in the code. In my opinion it's more important to explain why certain things where changed (e.g. why did did you have to make things `async`). No need to change anything in this PR, rather a suggestion for the future).
* Either add the label `ready-to-merge` to the PR so that the PR is automatically merged or press the green `merge` button yourself once all checks passed 🎉 (which I recommend for your first PR as pressing a button is way more fun and satisfying 😄 )

**Pull request merged**
